lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -0,0 +1,110 @@
+"""TODO: Add a proper docstring here.
+
+This is a placeholder docstring for this charm library. Docstrings are
+presented on Charmhub and updated whenever you push a new version of the
+library.
+
+Complete documentation about creating and documenting libraries can be found
+in the SDK docs at https://juju.is/docs/sdk/libraries.
+
+See `charmcraft publish-lib` and `charmcraft fetch-lib` for details of how to
+share and consume charm libraries. They serve to enhance collaboration
+between charmers. Use a charmer's libraries for classes that handle
+integration with their charm.
+
+Bear in mind that new revisions of the different major API versions (v0, v1,
+v2 etc) are maintained independently.  You can continue to update v0 and v1
+after you have pushed v3.
+
+Markdown is supported, following the CommonMark specification.
+"""
+
+import logging
+import functools
+import subprocess
+import yaml
+import base64
+import enum
+
+
+_logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "541fd767f90b40539cf7cd6e7db8fabf"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+def _snap(*args) -> str:
+    """Control snap by via executed `snap ...` commands.
+
+    Raises:
+        subprocess.CalledProcessError: Raised if snap command fails.
+    """
+    cmd = ["snap", *args]
+    _logger.debug("snap: Executing command %s", cmd)
+    try:
+        return subprocess.check_output(cmd, text=True)
+    except subprocess.CalledProcessError as e:
+        _logger.error("snap command failed: %s", e)
+        raise
+
+_get_config = functools.partial(_snap, "get", "slurm")
+_set_config = functools.partial(_snap, "set", "slurm")
+
+class Service(enum.Enum):
+    SLURMD = "slurmd"
+    SLURMCTLD = "slurmctld"
+    SLURMDBD = "slurmdbd"
+    SLURMRESTD = "slurmrestd"
+
+    @property
+    def config_name(self) -> str:
+        if self is Service.SLURMCTLD:
+            return "slurm"
+        return self.value
+
+class SlurmManager:
+    def __init__(self, service: Service):
+        self._service = service
+
+    def install(self):
+        # TODO: Pin slurm to the stable channel
+        _snap("install", "slurm", "--channel", "edge")
+
+    def start(self):
+        _snap("start", "--enable", f"slurm.{self._service.value}")
+
+    def restart(self):
+        _snap("restart", f"slurm.{self._service.value}")
```

Comment:
Thought: The original slurm manager uses different methods to restart slurm and munged, but maybe it would be better to have a single restart method for simplicity?

Reply:
No, I would keep them separate. We shouldn't have to restart the Slurm services when cycling out the current munge.key for a new one - some clusters have a key refresh policy, especially when introducing new partitions - and we shouldn't have to restart munge when making updates to the Slurm configuration. It introduces more potential points of failure imo.